client/tests/phpunit/Controller/OrganisationControllerTest.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace AppBundle\Controller;
+
+use AppBundle\Entity\Organisation;
+use AppBundle\Entity\User;
+use AppBundle\Model\Email;
+use AppBundle\Service\Mailer\MailFactory;
+use AppBundle\Service\Mailer\MailSender;
+use Prophecy\Argument;
+use Symfony\Component\HttpFoundation\Response;
+
+class OrganisationControllerTest extends AbstractControllerTestCase
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+    }
+
+    public function testAddAction(): void
+    {
+        $this->mockLoggedInUser(['ROLE_PROF_ADMIN']);
+
+        $emailAddress = 'invited@mailbox.example';
+        $user = (new User())
+            ->setId(21)
+            ->setEmail($emailAddress)
+            ->setRegistrationToken('invitation-token');
+
+        $organisation = (new Organisation())
+            ->setId(14)
+            ->setName('Test organisation');
+
+        $this->restClient->get('v2/organisation/14', 'Organisation')->shouldBeCalled()->willReturn($organisation);
+        $this->restClient->get("user/get-team-names-by-email/$emailAddress", 'User')->shouldBeCalled()->willReturn(new User());
+        $this->restClient->post('user', Argument::any(), ['org_team_add'], 'User')->shouldBeCalled()->willReturn($user);
+        $this->restClient->put('v2/organisation/14/user/21', '')->shouldBeCalled()->willReturn($user);
+
+        $this->injectProphecyService(MailSender::class, function($mailSender) use ($emailAddress) {
+            $mailSender->send(Argument::that(function ($email) use ($emailAddress) {
+                return $email instanceof Email
+                    && $email->getToEmail() === $emailAddress
+                    && $email->getTemplate() === MailFactory::INVITATION_TEMPLATE_ID
+                    && strpos($email->getParameters()['link'], "user/activate/invitation-token") !== false;
+            }), Argument::cetera())->shouldBeCalled()->willReturn();
+        });
```

Comment:
I guess this is the drawback of the method - its not super easy to see whats going on here. But I'll take it over not being able to test controllers!

Reply:
One option is to break it down into separate `Argument` calls rather than use a callback. I'm not convinced that's more readable though, because the language of Prophecy is a bit confusing.

```suggestion
        $this->injectProphecyService(MailSender::class, function($mailSender) use ($emailAddress) {
            $mailSender
                ->send(Argument::allOf(
                    Argument::type(Email::class),
                    Argument::which('getToEmail', $emailAddress),
                    Argument::which('getTemplate', MailFactory::INVITATION_TEMPLATE_ID),
                    Argument::that(function ($email) {
                        return strpos($email->getParameters()['link'], "user/activate/invitation-token") !== false;
                    })
                ), Argument::any())
                ->shouldBeCalled()
                ->willReturn();
        });
```

That said, maybe I could just lay it out a bit clearer:
```suggestion
        $mailSender = $this->injectProphecyService(MailSender::class);
        $mailSender
            ->send(Argument::that(function ($email) use ($emailAddress) {
                return $email instanceof Email
                    && $email->getToEmail() === $emailAddress
                    && $email->getTemplate() === MailFactory::INVITATION_TEMPLATE_ID
                    && strpos($email->getParameters()['link'], "user/activate/invitation-token") !== false;
            }), Argument::any())
            ->shouldBeCalled()
            ->willReturn();
```